File: backend/app/schemas/hotel.py
from datetime import time
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator


class ORMModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class HotelOut(ORMModel):
    id: UUID
    name: str
    slug: str
    logo_object_key: str | None
    address_line1: str | None
    address_line2: str | None
    city: str | None
    state: str | None
    country: str
    postal_code: str | None
    phone: str | None
    email: str | None
    website: str | None
    description: str | None
    timezone: str
    status: str


class ServiceItemOut(ORMModel):
    id: UUID
    name: str
    price: Decimal
    is_active: bool


class ServiceItemCreate(BaseModel):
    name: str = Field(min_length=2, max_length=120)
    price: Decimal = Field(ge=0)


class ServiceItemUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=120)
    price: Decimal | None = Field(default=None, ge=0)
    is_active: bool | None = None


class HotelUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=200)
    address_line1: str | None = Field(default=None, max_length=255)
    address_line2: str | None = Field(default=None, max_length=255)
    city: str | None = Field(default=None, max_length=120)
    state: str | None = Field(default=None, max_length=120)
    country: str | None = Field(default=None, max_length=120)
    postal_code: str | None = Field(default=None, max_length=32)
    phone: str | None = Field(default=None, max_length=32)
    email: EmailStr | None = None
    website: str | None = Field(default=None, max_length=255)
    description: str | None = Field(default=None, max_length=4000)
    timezone: str | None = Field(default=None, max_length=64)


class HotelSettingsOut(ORMModel):
    check_in_time: time
    check_out_time: time
    cancellation_policy: str | None
    no_show_policy: str | None
    invoice_prefix: str
    invoice_next_number: int
    booking_prefix: str
    booking_next_number: int
    tax_inclusive_pricing: bool
    currency: str
    early_checkin_grace_minutes: int
    late_checkout_grace_minutes: int
    access_mode: str = "full"


class HotelSettingsUpdate(BaseModel):
    check_in_time: time | None = None
    check_out_time: time | None = None
    cancellation_policy: str | None = Field(default=None, max_length=4000)
    no_show_policy: str | None = Field(default=None, max_length=4000)
    invoice_prefix: str | None = Field(default=None, min_length=1, max_length=32)
    booking_prefix: str | None = Field(default=None, min_length=1, max_length=32)
    tax_inclusive_pricing: bool | None = None
    early_checkin_grace_minutes: int | None = Field(default=None, ge=0, le=720)
    late_checkout_grace_minutes: int | None = Field(default=None, ge=0, le=720)


class GstSettingsOut(ORMModel):
    is_gst_registered: bool
    gstin: str | None
    legal_name: str | None
    trade_name: str | None
    address: str | None
    state: str | None
    state_code: str | None
    default_cgst_rate: Decimal
    default_sgst_rate: Decimal
    default_igst_rate: Decimal
    version: int


class GstSettingsUpdate(BaseModel):
    is_gst_registered: bool | None = None
    gstin: str | None = Field(default=None, min_length=15, max_length=15)
    legal_name: str | None = Field(default=None, max_length=255)
    trade_name: str | None = Field(default=None, max_length=255)
    address: str | None = Field(default=None, max_length=2000)
    state: str | None = Field(default=None, max_length=120)
    state_code: str | None = Field(default=None, min_length=2, max_length=2)
    default_cgst_rate: Decimal | None = Field(default=None, ge=0, le=50)
    default_sgst_rate: Decimal | None = Field(default=None, ge=0, le=50)
    default_igst_rate: Decimal | None = Field(default=None, ge=0, le=100)

    @field_validator("gstin")
    @classmethod
    def validate_gstin(cls, value: str | None) -> str | None:
        if value is None:
            return value
        value = value.upper().strip()
        import re

        # 15 chars: 2-digit state code, 10-char PAN, entity code, 'Z', checksum
        if not re.fullmatch(r"\d{2}[A-Z]{5}\d{4}[A-Z][A-Z\d]Z[A-Z\d]", value):
            raise ValueError("Invalid GSTIN format")
        return value

    @field_validator("state_code")
    @classmethod
    def validate_state_code(cls, value: str | None) -> str | None:
        if value is None:
            return value
        if not value.isdigit():
            raise ValueError("State code must be two digits")
        return value


class PaymentConfigOut(BaseModel):
    """Config view for Owner/Manager/Admin — includes the raw UPI ID."""

    upi_id: str | None
    config_version: int
    has_logo: bool
    qr_version: int


class PaymentConfigUpdate(BaseModel):
    upi_id: str = Field(min_length=3, max_length=256)

    @field_validator("upi_id")
    @classmethod
    def validate_upi(cls, value: str) -> str:
        import re

        value = value.strip()
        # handle@psp — conservative UPI VPA validation
        if not re.fullmatch(r"[a-zA-Z0-9._-]{2,}@[a-zA-Z]{2,64}", value):
            raise ValueError("Invalid UPI ID format (expected handle@psp)")
        return value


class PaymentQrOut(BaseModel):
    """Worker-safe response — never contains the raw UPI ID."""

    qr_available: bool
    qr_version: int
    payment_label: str

File: frontend/src/types/hotel.ts
export interface HotelOut {
  id: string;
  name: string;
  slug: string;
  logo_object_key: string | null;
  address_line1: string | null;
  address_line2: string | null;
  city: string | null;
  state: string | null;
  country: string;
  postal_code: string | null;
  phone: string | null;
  email: string | null;
  website: string | null;
  description: string | null;
  timezone: string;
  status: string;
  total_rooms?: number | null;
  map_id?: string | null;
  /** Staff-attendance geofence (client 09/2026) */
  geofence_enabled?: boolean;
  latitude?: string | null;
  longitude?: string | null;
  geofence_radius_m?: number;
  attendance_grace_minutes?: number;
}

export interface HotelSettingsOut {
  check_in_time: string;
  check_out_time: string;
  cancellation_policy: string | null;
  no_show_policy: string | null;
  invoice_prefix: string;
  invoice_next_number: number;
  booking_prefix: string;
  booking_next_number: number;
  tax_inclusive_pricing: boolean;
  currency: string;
  early_checkin_grace_minutes: number;
  late_checkout_grace_minutes: number;
  early_checkin_fee_per_hour: string;
  late_checkout_fee_per_hour: string;
  access_mode: "full" | "checkin_expense" | "checkin_only";
  collect_emergency_contact: boolean;
  collect_vehicle_details: boolean;
  /** "Powered by DigitalMyHotels" invoice branding (plan §3.8). */
  show_powered_by: boolean;
}

export interface HotelImageOut {
  id: string;
  position: number;
}

export interface GstSettingsOut {
  /** Client 09/2026 GST modes:
   *  - "no_gst": GST flow hidden everywhere in the partner system
   *  - "included_by_hotel": GST inside the price, never shown to customer
   *  - "included_by_customer": GST added on top and displayed */
  gst_mode: "no_gst" | "included_by_hotel" | "included_by_customer";
  is_gst_registered: boolean;
  gstin: string | null;
  legal_name: string | null;
  trade_name: string | null;
  address: string | null;
  state: string | null;
  state_code: string | null;
  default_cgst_rate: string;
  default_sgst_rate: string;
  default_igst_rate: string;
  version: number;
}

export interface PaymentConfigOut {
  upi_id: string | null;
  config_version: number;
  has_logo: boolean;
  qr_version: number;
}

export interface PaymentQrOut {
  qr_available: boolean;
  qr_version: number;
  payment_label: string;
}

export interface RoomTypeOut {
  id: string;
  code: string;
  name: string;
  description: string | null;
  base_price: string;
  extra_guest_price: string;
  hourly_rate: string | null;
  max_occupancy: number;
  is_active: boolean;
}

export interface RoomOut {
  id: string;
  room_number: string;
  floor: string | null;
  bed_type: string | null;
  max_adults: number | null;
  max_children: number | null;
  status: RoomStatus;
  is_active: boolean;
  notes: string | null;
  room_type_id: string;
  room_type_name: string | null;
  amenities: string[];
  // ── Derived reservation context (room-status redesign, 15/09) ─────────────
  /** A confirmed booking's stay window includes today (guest expected). */
  arriving_today: boolean;
  /** "HH:MM" expected arrival time for today's booking. */
  arrival_time: string | null;
  /** ISO date of the earliest FUTURE confirmed booking on this room. */
  next_booking_date: string | null;
  /** "HH:MM" check-in time of that future booking. */
  next_booking_time: string | null;
  /** Current in-house guest checks out today. */
  departing_today: boolean;
  /** "HH:MM" expected checkout time today. */
  departure_time: string | null;
}

// ── Date-aware availability ───────────────────────────────────────────────────

export interface RoomAvailableItem {
  id: string;
  room_number: string;
  floor: string | null;
  bed_type: string | null;
  status: RoomStatus;
  is_active: boolean;
  room_type_id: string;
  room_type_name: string | null;
  room_type_base_price: string;
  room_type_hourly_rate: string | null;
  max_occupancy: number;
  amenities: string[];
  /** For currently-occupied rooms: ISO date of current guest's checkout. */
  current_checkout_date: string | null;
  /** For currently-occupied rooms: "HH:MM" checkout time (hotel local time). */
  current_checkout_time: string | null;
  /** Next confirmed booking starting on/after the requested checkout —
      "Booked from Sep 24, 14:00 — free for your dates". */
  next_booking_date: string | null;
  /** "HH:MM" check-in time of that next booking. */
  next_booking_time: string | null;
}

export type UnavailableReason =
  | "booked"
  | "occupied"
  | "cleaning"
  | "maintenance"
  | "out_of_service";

export interface RoomUnavailableItem extends RoomAvailableItem {
  unavailable_reason: UnavailableReason;
  /** ISO date string — when this room will next be free (for booked rooms). */
  occupied_until: string | null;
  /** "HH:MM" checkout time on the occupied_until date (hotel local). */
  occupied_until_time: string | null;
  overlapping_booking_count: number;
}

export interface RoomAvailabilityOut {
  check_in_date: string;
  check_out_date: string;
  available: RoomAvailableItem[];
  unavailable: RoomUnavailableItem[];
  total_rooms: number;
}

export type RoomStatus =
  | "available"
  | "reserved"
  | "occupied"
  | "cleaning_required"
  | "cleaning_in_progress"
  | "clean_ready"
  | "inspection_required"
  | "maintenance"
  | "out_of_service";

export interface RoomStatusSummaryOut {
  total: number;
  counts: Partial<Record<RoomStatus, number>>;
}

export interface TeamMemberOut {
  membership_id: string;
  user_id: string;
  full_name: string;
  email: string;
  phone: string | null;
  role_code: string;
  role_name: string;
  status: string;
  is_active: boolean;
  last_login_at: string | null;
}

export interface ListOut<T> {
  items: T[];
  total: number;
}
